File: Common/TFEGameController.m
//
//  TFEGameController.m
//  TwentyFortyEight
//
//  Created by Joshua Caswell on 9/29/16.
//  Copyright © 2016 Josh Caswell. All rights reserved.
//

#import "TFEGameController.h"
#import "TFEGameController+Private.h"
#import "TFEMainScene.h"
#import "TFEBoard.h"
#import "TFELabel.h"

@implementation TFEGameController
{
    uint32_t _score;
}

- (instancetype)init
{
    self = [super initWithNibName:@"MainView" bundle:nil];
    if( !self ) return nil;
    
    _scene = [TFEMainScene sceneWithSize:CGSizeMake(100, 100)];
    _board = [TFEBoard boardWithController:self
                                     scene:_scene];
    
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    [[self scene] setSize:[[self gameView] bounds].size];
}

- (void)viewWillAppear
{
    [self viewWillAppear:NO];
}

- (void)viewWillAppear:(BOOL)animated
{
    [[self gameView] presentScene:[self scene]];
}

- (BOOL)becomeFirstResponder
{
    return YES;
}

//MARK: Board communication

- (void)gameDidEndInVictory:(BOOL)victorious
{
    [[self scene] gameDidEndInVictory:victorious];
}

- (void)updateScoreTo:(uint32_t)new_score
{
    [[self scoreLabel] setText:[NSString stringWithFormat:@"%d", new_score]];
}

@end

File: Common/TFEGameController.swift
//
//  TFEGameController.swift
//  TwentyFortyEight
//
//  Created by Joshua Caswell on 10/1/16.
//  Copyright © 2016 Josh Caswell. All rights reserved.
//

import SpriteKit

class TFEGameController : TFEViewController
{
    @IBOutlet weak var gameView: SKView!
    @IBOutlet weak var scoreLabel: TFELabel!
    
    let scene: TFEMainScene
    private(set) var board: TFEBoard!
    
    private var _score: UInt32 = 0
    
    init()
    {
        self.scene = TFEMainScene(size: CGSizeMake(100, 100))
        
        #if os(OSX)
        super.init(nibName: "MainView", bundle: nil)!
        #elseif os(iOS)
        super.init(nibName: "MainView", bundle: nil)
        #endif
        
        self.board = TFEBoard(controller: self, scene: self.scene)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) is unimplemented")
    }
    
    override func viewDidLoad()
    {
        super.viewDidLoad()
        
        self.scene.size = self.gameView.bounds.size
    }
    
    override func viewWillAppear(animated: Bool)
    {
        self.gameView.presentScene(self.scene)
    }
    
    override func becomeFirstResponder() -> Bool
    {
        return true
    }
    
    // Message sent to self from platform-specific user input extensions.
    // This is necessary because `board` is not mutable inside the extensions, and `moveNodes(inDirection:)` is mutating.
    func userDidInput(direction: TFENodeDirection)
    {
        self.board.moveNodes(inDirection: direction)
    }
    
    //MARK: - Communication from board
    
    func gameDidEnd(inVictory victorious: Bool)
    {
        self.scene.gameDidEnd(inVictory: victorious)
    }
    
    func updateScore(to score: UInt32)
    {
        self.scoreLabel.text = String(format: "%d", score)
    }
}
